### PR TITLE
Track a distinct compile cache for julia-debug.

### PR DESCRIPTION
**Pull request by twadleigh**
Addresses #33668.

### CODE REVIEW

File: base/loading.jl [hunk 1]
```diff
@@ -633,7 +633,7 @@ end
 
 cache_file_entry(pkg::PkgId) = joinpath(
     "compiled",
-    "v$(VERSION.major).$(VERSION.minor)",
+    "v$(VERSION.major).$(VERSION.minor)" * (JLOptions().debug_level > 1 ? "-debug" : ""),
```
Comment (twadleigh) [inline base/loading.jl]: `JLOptions().debug_level > 1` is my best guess for the most robust way to test for `julia-debug`, but others much more knowledgeable may have a different opinion.

Comment (twadleigh) [reply base/loading.jl]: Specifically, I chose to prefer it to something like:
```
endswith(basename(unsafe_string(Base.JLOptions().julia_bin)),"-debug")
```
so as to be insensitive to a name change of the executable.

**Comment by tkf** [comment]
FYI an alternative implementation may be to include `JLOptions().julia_bin` in `package_slug`.  I implemented this in #29914:

https://github.com/JuliaLang/julia/blob/fa5f1b178fa80bf31d6723a7d734b296e990cc06/base/loading.jl#L136-L141

A major advantage might be that using `JLOptions().julia_bin` (and `.image_file`) would support differentiating arbitrary binaries which are not precompile-cache -compatible.  A disadvantage is that renaming the binary would create a different set of precompile-cache files.  However, this can be avoided by using symlinking `julia`/`julia-debug` binary.

**Comment by twadleigh** [comment]
@tkf, different system images similarly clobber the compile cache? Tweaking the slug is one way to go. You might also consider qualifying the path, like I did here. It might be a more palatable solution to the maintainers, as it is guaranteed not to interact with the existing implementation.

**Comment by tkf** [comment]
> different system images similarly clobber the compile cache?

Yes, especially because you can use PackageCompiler to bundle a different set of packages in the system image.

> You might also consider qualifying the path

Yes, using suffix as you do in this PR does have a benefit like refreshing precompilation cache is as easy as `rm -r ~/.julia/compiled/v1.x-debug` and does not slow down normal `julia` usage.

**Comment by twadleigh** [comment]
@tkf, I think I'll attempt a modest extension of this solution that would, for instance, keep the caches distinct for distinct build images.

**Comment by tkf** [comment]
@twadleigh I actually don't know how much core devs want to play with it.  So maybe just wait for a response?  I don't want to request a feature that would be rejected by core devs :)

**Comment by twadleigh** [comment]
@tkf, fair enough. I'll propose an extension in a separate PR only if/when this is merged.

**Comment by tkf** [comment]
FYI, @KristofferC just merged PR #29914.  I'm not sure @KristofferC was aware of this patch.  Sorry, I should have mentioned there that an alternative patch was suggested in this PR...

File: base/loading.jl (hunk 1 above)
Comment (vchuravy) [reply base/loading.jl]: `JLOptions().debug_level` corresponds to the `-g` option and not whether the binary was build with debug information embedded

Comment (twadleigh) [reply base/loading.jl]: Yes, the test should be `ccall(:jl_is_debugbuild, Cint, ()) > 0`.

**Comment by twadleigh** [comment]
@tkf, I'll close this. If it turns out master still doesn't have the functionality I'm looking for, I'll reopen.

**Pull request closed without merging**